### PR TITLE
Monitoring VMI Phase Transition Times

### PR DESCRIPTION
**Pull request by davidvossel**
This PR introduces a new VMI status field that tracks phase transition timestamps. It also adds two new histogram metrics to help track how long it takes for phase transitions to occur. 

The new histogram metric `kubevirt_vmi_phase_transition_time_from_creation_seconds` tracks now long since creation it took to transition to the current phase a VMI is in. This would be useful if you wanted to track how long it takes for VMI's to go from creation to `Running`. 

Here's an example query

```
histogram_quantile(0.95, sum(rate(kubevirt_vmi_phase_transition_time_from_creation_seconds_bucket{phase="Running"}[5m])) by (le))
```

Using variations of that query, we can create grafana graphs that indicate how long it takes for VMIs to reach running, and which phase the VMI spends the most time in.

histogram-median-p95

A similar histogram called `kubevirt_vmi_phase_transition_time_seconds` has been introduced that tracks the time between the last_phase and current phase of a vmi. This would be useful for targetting how long it takes to go from schedulling to scheduled, or scheduled to running.

```release-note
Addition of kubevirt_vmi_phase_transition_seconds_since_creation to monitor how long it takes to transition a VMI to a specific phase from creation time.
```

### CODE REVIEW

**Comment by davidvossel** [comment]
related to #5732. this helps track the time spent in different phases.

File: pkg/monitoring/vms/prometheus/prometheus.go
Comment (davidvossel) [reply pkg/monitoring/vms/prometheus/prometheus.go:554]: > Is a ConstMetric a standalone metric? In other words, when prometheus scrapes this will it get a time for Pending and Scheduling or will it only see the most recent transition time and phase?

What's reported is the most recent transition time for a VMI at the point in time the metric is scraped. There are labels to this metric that allow selection by name, namespace, and phase.

So, Taking the average of `kubevirt_vmi_phase_transition_seconds_since_creation{phase=Running}` at any point in time would give you an indication of the average amount of time all currently running VMIs took to reach Running.  This could further be filtered by namespace and name. So if you started VMIs in different batches and wanted to separate the results, launching each batch in a separate namespace would be one way of doing that.

Comment (rthallisey) [reply pkg/monitoring/vms/prometheus/prometheus.go:554]: The reason I'm asking is the time to running doesn't always tell the whole story.  For instance, if my PVCs were taking a long time to create for whatever reason, I would see a huge chunk of time spent in Pending.  But the `...{phase=Running}` metric doesn't always clue me into what the problem is, only that my time to running is a little longer than expected.  It sounds like though you can filter for `kubevirt_vmi_phase_transition_seconds_since_creation{phase=Pending}` and get that data, however we would be dependent on how often prometheus scrapes e.g. 60s interval for a scrape and VMs spend 59s in Pending.  I'm wondering if we can get a more granular look at the transition time.

Comment (rmohr) [reply pkg/monitoring/vms/prometheus/prometheus.go:554]: > The reason I'm asking is the time to running doesn't always tell the whole story. For instance, if my PVCs were taking a long time to create for whatever reason, I would see a huge chunk of time spent in Pending. But the ...{phase=Running} metric doesn't always clue me into what the problem is, only that my time to running is a little longer than expected. 

In prow I decide to record `last_phase` and `phase` https://github.com/kubernetes/test-infra/pull/14695 in the label to immediately indicate where it takes the time. It is then more verbose to get a sum of sub-states, but you can easily see where most of the time is spent and you can still get the overall-time instantly.

Either way I would suggest using a histogram and I would lean towards having `last_phast` and `phase` in the metrics.

> It sounds like though you can filter for kubevirt_vmi_phase_transition_seconds_since_creation{phase=Pending} and get that data, however we would be dependent on how often prometheus scrapes e.g. 60s interval for a scrape and VMs spend 59s in Pending. I'm wondering if we can get a more granular look at the transition time.

Two things on that:

  * In general I would be less interested in following individual VMIs. I would mostly want to see the histograms and compare the histogram with previous runs of the tests where e.g. 100VMs were started in parallel with well known VMI specs.
 * For specific VMI tracking: At mimimum for performance testing we should allow setting the VMI name/namespace on the metric. You will get the data for every transition if you choose the test-timing right. Basically something like this: Wait for a minute, add a unique test label to each metric on the prometheus config level, run the tests, wait for another minute. You will get all the data. When you then do queries (including sum, substractions, ...), you just have to choose the time-frame bigger than the test-window.
  * If we have long-running sub-tasks (like pvc provisioning) I think it would be worth making them explicitly visible in the metric.

Comment (rthallisey) [reply pkg/monitoring/vms/prometheus/prometheus.go:554]: >Either way I would suggest using a histogram and I would lean towards having last_phast and phase in the metrics
>
Adding `last_phase` as a label sounds like it would help, however I still don't think that get us every detail.  I'm expecting a metric that has as its keys the VMs phase and Name, so the metric will have a record of each phase per VM.  I was thinking something like a [GaugeVec](https://pkg.go.dev/github.com/m3db/client_golang/prometheus#GaugeVec) (or [HistogramVec](https://pkg.go.dev/github.com/m3db/client_golang/prometheus#HistogramVec)).  Might look something like:

```go
	transition := prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Namespace: ns,
			Subsystem: "phase",
			Name:      "transitionTime",
			Help:      "Phase transition time per VM.",
		},
		[]string{
			// Which VM we're measuring
			"vmName",
			// Which phase we measured
			"phase",
		},
	)
```

In Prometheus this should come out like:
```
phase_transitionTime_gauge{namespace=“perf-test”, phase=“Pending”, vmName=“perf-1”}                 59s
phase_transitionTime_gauge{namespace=“perf-test”, phase=“Scheduling”, vmName=“perf-1”}              60s
phase_transitionTime_gauge{namespace=“perf-test”, phase=“Running”, vmName=“perf-1”}                 75s
```

**Review by marceloamaral — CHANGES_REQUESTED** [review]

File: pkg/monitoring/vms/prometheus/prometheus.go
Comment (davidvossel) [reply pkg/monitoring/vms/prometheus/prometheus.go:554]: +1 for histogram, i'll change this.

Comment (davidvossel) [reply pkg/monitoring/vms/prometheus/prometheus.go:554]: I get the value of changing this metric to a histogram.  I'm unsure what tracking "last_phase" helps with though. 

Why wouldn't measuring each phase from creation time and putting that in a histogram expose the same data?

Comment (davidvossel) [reply pkg/monitoring/vms/prometheus/prometheus.go:554]: >  however we would be dependent on how often prometheus scrapes e.g. 60s interval for a scrape and VMs spend 59s in Pending. I'm wondering if we can get a more granular look at the transition time.

@rthallisey I can avoid this by keeping a list of all phase transitions on the VMI status and reporting that in the histogram data.

When it comes to tracking the time, what do you think is most useful? Tracking each phase's duration anchored to creation time, or tracking each phase's duration anchored to the previous phases transition time?

Comment (rthallisey) [reply pkg/monitoring/vms/prometheus/prometheus.go:554]: If we do time spent in each phase, then the server does the math and reports the result.  If we do time since create, then the client does the math.  I don't think either option closes off any use cases, so I like anchoring to the previous phase for convenience.

The one concern I have with anchoring to the previous phase is if the previous phase is missed.  For instance, can an edge triggered phase like `Scheduled` be missed?

>@rthallisey I can avoid this by keeping a list of all phase transitions on the VMI status and reporting that in the histogram data.
>
👍

Comment (marceloamaral) [reply pkg/monitoring/vms/prometheus/prometheus.go:554]: The advantage of having the interval between phases (`current_phase` - `last_phase`) is to apply aggregation (e.g. sum) in the metric.
If all phases have absolute time since creation, it will wrongly aggregate the time of all phases....

For example : 
`histogram_quantile(0.99,
   sum(irate(phase_transition_duration_miliseconds_bucket[1m]))
   by (vmName, le))
`

Otherwise, the query can become very complex to subtract the times....

Comment (davidvossel) [reply pkg/monitoring/vms/prometheus/prometheus.go:554]: > The one concern I have with anchoring to the previous phase is if the previous phase is missed. For instance, can an edge triggered phase like Scheduled be missed?

yes it is possible. 

Ultimately what I want is two things

1. I want to know how long it takes until a VMI reaches running.
2. I want to know what phase we're spending the most time in.

I think the best way to solve this is two metrics. Keep what I have with creation to running, and add a histogram for transition times between phases as well.

Then we get everything. Someone can create a graph measuring just the time until Running and break down where we spend the most time in what phase.

**Comment by davidvossel** [comment]
> I also think that a histogram is more interesting to analyze in this metric.
> For example, after creating a large amount of VMs, the 90 and 99-percentile are much more useful than average...

@marceloamaral we don't have to use a histogram to get percentiles. For example...

```quantile(0.95, kubevirt_vmi_phase_transition_seconds_since_creation{phase="Running"})```

that gives me a graph of the p95 for long it takes VMIs to becoming running in seconds. It's pretty neat. I can have p95 for each phase and overlay them all to see where we spend the most time. That's without the histogram.


@rmohr @rthallisey I added a histogram to this PR. I question it's usefulness though. It works for getting some detailed per phase transition times, but in many ways doesn't give me as accurate of a picture as the gauge that measures phase to creation time.

Do you all think I should leave the histogram in, or is the gauge good enough?

**Comment by rthallisey** [comment]
>Do you all think I should leave the histogram in, or is the gauge good enough?
>
Have a screenshot?  I'm a little more familiar with gauges and I think that would work.  Not so much with histograms.

**Comment by marceloamaral** [comment]
@davidvossel, there are some very important differences between those approaches, which in fact are not very clear.... 
Let me try to explain it... Btw, sorry for the long explanation, I hope it will shed some light on the differences... 

Prometheus has two different **`aggregation operators quantile`** `Summary` and `Histogram`. `Histogram` is the evolution of `Summary` and they differ. For example:
> Summary: `request_duration_summary{quantile="0.95"}`
> Histogram: `histogram_quantile(0.95, http_request_duration_seconds_bucket) by (le))`

The key difference between `Summary` and `Histogram` is: in `Summary` the quantile is calculated on the client-side, and in the `Histogram` the calculation is on the server (Prometheus) side. Since the latter, the calculation is in the server, it is more flexible. For example, the `Summary` will have hard-coded the supported percentile (e.g.,  0.95). Additionally, [the Prometheus official documentation](https://prometheus.io/docs/practices/histograms/) says: "aggregating the precomputed quantiles from a `Summary` rarely makes sense. In this particular case, averaging the quantiles yields statistically nonsensical values." Then:
> avg(http_request_duration_seconds{quantile="0.95"}) // **BAD!**
> histogram_quantile(0.95, sum(rate(http_request_duration_seconds_bucket[5m])) by (le))  // **GOOD.**

That is, we cannot sum the phases' latency in the `Summary`.... Also, in addition to being aggregatable, `Histogram` is cheaper on the client too as counters are fast to increment.

Finally, note that **`aggregation operators quantile`** should not be confused with the **`function quantile_over_time`**... 
The `quantile_over_time` ["calculates quantiles from sample values that already exist as a time series in Prometheus, i.e. where the sample values themselves are treated as observations"](https://docs.google.com/document/d/1uSenXRDjDaJLV3qnSD09GqgPdEEDPjER0mVsnGaCYF0/edit#heading=h.vuk4s3dgrhjr). Which is something completely different and does not represent the quantile of client samples of the metric. To better understand the difference between `aggregation operators quantile` and `function quantile_over_time` on top of a `Gauge` metric, as you suggested, we need first to understand how `Histogram` and `Gauges` are calculated and scraped. 

On one hand, `Gauges` can go up and down over time, and scrapes take a snapshot of the current value. So, if within the interval between scrapes, the samples varied a lot, the `Gauges` will only report the last value (which can be low or high)...
`Summary` with quantile and `Histogram` both create a "histogram" in the client (in the code) computing all samples of the data. Then, when Prometheus scrapes those metrics from the client, it will get the incremental data of the "histogram", representing the latency variation in all samples collected between scrapes.

Therefore, the `quantile_over_time` on top of a `Gauge` metric, will report something completely different from the quantile of a `Histogram`. Not that, this is not only a matter of granularity of the metric but the accuracy, the concept is very different.

**Comment by marceloamaral** [comment]
A good practice to have a latency metric is to use `Histogram` instead of  `Gauge`, since `Gauge` sometimes does not show useful data for latency, as we discussed before...

You can see that most of the Kubernetes latency metrics are using `Histogram` (e.g., `apiserver_request_duration_seconds_bucket`), and most of the `Gauge` metrics that have seconds in its name are not calculating latency but usage (e.g., `container_cpu_usage_seconds_total`).

So, IMHO, in the case we need to select between keep `Histogram` or `Gauge` metrics here, we should go for `Histogram`
Actually, I think we don't even need the `Gauge` metrics for latency...

**Comment by davidvossel** [comment]
> Have a screenshot? I'm a little more familiar with gauges and I think that would work. Not so much with histograms.

yeah. 

p95-running

There i run a load test where i incrementally introduce more and more VMIs, we can see the p95 of the running time increase as the load test increases which indicates that the worst case (time from creation to running) increases as load increases.

**Comment by marceloamaral** [comment]
>The histogram doesn't give us this. It instead gives us something similar to how someone would want to visualize http response times where phase transition times are the equivalent of "http" requests. Maybe someone finds that useful. I don't for performance testing purposes.

Thanks David! 

Sorry, I am not sure If I understood what you mean about histogram is not suitable.

Histogram is highly used for performance tests to calculate percentiles on a range of samples, specially for latency....

For example, `histogram_quantile(0.95, sum(rate(phase_duration_seconds_bucket{phase="Running"}[5m])) by (le))` would show some similar graph that you showed, but the quantile will be applied for all the VM creation samples, instead of only in the sample of prometheus observations. Since `Gauges` will only report the last value.

**Comment by rmohr** [comment]
> For example, `histogram_quantile(0.95, sum(rate(phase_duration_seconds_bucket{phase="Running"}[5m])) by (le))` would show some similar graph that you showed, but the quantile will be applied for all the VM creation samples, instead of only in the sample of prometheus observations. Since `Gauges` will only report the last value.

I think the trick why this works for David is that he has a gauge for every VM for every phase. The gauge basically never goes down. For as long as he keeps the VMs running, the gauges with a constant value will be reported.

However, I disagree with

> this is exactly what I wanted to see. The histogram doesn't give us this. It instead gives us something similar to how someone would want to visualize http response times where phase transition times are the equivalent of "http" requests. Maybe someone finds that useful. I don't for performance testing purposes.

too. I would definitely want heatmaps per phase. In addition I can still calculate something like David showed based on the included _sum and _count metrics which are part of the histogram. In addition we don't need vm and namespace labels on the metrics to still get metrics for all VMs which makes it cheap and it can also easily be enabled in production.

**Comment by davidvossel** [comment]
The PR has been updated as well as the description. The code consists of two histograms now. One that tracks each phase from creation, the other that tracks each phase with the last phase.

File: pkg/monitoring/vmistats/collector.go
Comment (marceloamaral) [inline pkg/monitoring/vmistats/collector.go:242]: I would suggest to also have a millisecond precision in the histogram.
There are 12 buckets now, but we could make it 15...
`(0.1 * time.Second.Seconds()),`
`(0.5 * time.Second.Seconds()),`
Or just the 0.5....

**Comment by rthallisey** [comment]
What do you think about using ` pkg/monitoring/perf` and `perf` for the package name?  Then I can see all worker queue metrics landing in `perf` and later all the scale metrics land in ` pkg/monitoring/scale`.

File: pkg/monitoring/vmistats/collector.go
Comment (rmohr) [inline pkg/monitoring/vmistats/collector.go:62]: I think this is a left-over.

**Review by rmohr — CHANGES_REQUESTED** [review]
One left-over I think. The rest looks good to me.

**Comment by davidvossel** [comment]
> What do you think about using pkg/monitoring/perf and perf for the package name? Then I can see all worker queue metrics landing in perf and later all the scale metrics land in pkg/monitoring/scale.

@rthallisey I'm fine with a new `perf` package for these sorts of metrics.  What would be the difference between what you'd want to see in perf vs a scale package though?

File: pkg/monitoring/vmistats/collector.go
Comment (davidvossel) [reply pkg/monitoring/vmistats/collector.go:62]: oh yeah... thanks

**Comment by rthallisey** [comment]
I was thinking `scale` would have metrics that record the number of APIs calls Kubevirt makes to Kubernetes.  That was the initial one I was thinking of, but there could be more.

File: pkg/monitoring/vmistats/collector.go
Comment (davidvossel) [reply pkg/monitoring/vmistats/collector.go:234]: updated, i added a .5 and then some more fine granular buckets in the middle. < .5 isn't that useful to us i don't think.

File: pkg/monitoring/vmistats/collector.go
Comment (davidvossel) [reply pkg/monitoring/vmistats/collector.go]: fixed, thanks

**Comment by davidvossel** [comment]
> I was thinking scale would have metrics that record the number of APIs calls Kubevirt makes to Kubernetes. That was the initial one I was thinking of, but there could be more.

i think a general package just aimed at perf/scale could probably encompass all of this. What would you think about starting with a single package called perf-scale, which would be for metrics related to measuring control plane behavior

**Comment by rthallisey** [comment]
>i think a general package just aimed at perf/scale could probably encompass all of this. What would you think about starting with a single package called perf-scale, which would be for metrics related to measuring control plane behavior
>
Works for me.  Probably easier to start that way and we can split it up if we find they need to be independent.

**Comment by rmohr** [comment]
/lgtm

Will approve once @rthallisey  and @marceloamaral had a chance to look at the PR.

**Comment by marceloamaral** [comment]
Everything looks good to me, I think we can approve this!

File: pkg/monitoring/vmistats/collector.go
Comment (rthallisey) [inline pkg/monitoring/vmistats/collector.go:206]: We can miss edge triggered events like `Scheduled` and that can misrepresent the transition from `Scheduling -> Running`.  It can also misrepresent `Scheduled`, if the majority of VMs were quick and one was slow so the metrics only pickup the slow `Scheduled`.  What do you think about `Scheduled` being merged into `Scheduling -> Running`?

File: pkg/monitoring/vmistats/collector_test.go
Comment (rthallisey) [inline pkg/monitoring/vmistats/collector_test.go:20]: perfscale

File: pkg/monitoring/vmistats/collector.go
Comment (davidvossel) [reply pkg/monitoring/vmistats/collector.go:206]: > We can miss edge triggered events like Scheduled and that can misrepresent the transition from Scheduling -> Running

I'm not sure that's possible. 

We're not updating the histogram based on a scrape interval. We're actually watching all VMI's and updating the histogram client side from the informer.

So the histogram gets continually updated independently of the scrape interval, and the scrape interval is getting an update of all the new bucket counts, totals, sums, etc... At least that's my understanding

File: pkg/monitoring/vmistats/collector_test.go
Comment (davidvossel) [reply pkg/monitoring/vmistats/collector_test.go]: yup, made the change.

The package is called perfscale, and it contains these two new histograms.

File: pkg/monitoring/vmistats/collector.go
Comment (rthallisey) [reply pkg/monitoring/vmistats/collector.go]: >We're actually watching all VMI's and updating the histogram client side from the informer.
>
That's what I'm referring to.  

>We can miss edge triggered events like Scheduled
>
Went back to look at the reconcile code..., `Scheduled` isn't actually an edge triggered event like I thought because of the `-ed` postfix (other examples: Failed, Succeeded, Deleted).  The phase 'Scheduling' is when the virt-controller is waiting for the virt-launcher pod to be ready.  And `Scheduled` is actually a level triggered event where the virt-handler has taken over for reconciling the VMI.  `Scheduling` + `Scheduled` is really `Scheduling`, but I see the value in recording the transition.

The reason why I thought `Scheduled` phase would sometimes be missed is because I thought it was a race between libvirt noticing guest activity and the handler processing that the pod is Running. 

We won't miss any phases.

**Comment by rthallisey** [comment]
lgtm

**Comment by marceloamaral** [comment]
It seems that we all agreed to approve this PR.
/approve
/retest

File: pkg/monitoring/perfscale/perfscale_suite_test.go
Comment (rmohr) [inline pkg/monitoring/perfscale/perfscale_suite_test.go:10]: @davidvossel this needs to be changed to

```
func TestPerfscale(t *testing.T) {
   testutils.KubeVirtTestSuiteSetup(t)
}
```

to make the unittests pass.

**Review by rmohr — COMMENTED** [review]
a minor issue needs to be address to make the unit tests pass.

/hold

File: pkg/monitoring/perfscale/perfscale_suite_test.go
Comment (davidvossel) [reply pkg/monitoring/perfscale/perfscale_suite_test.go:9]: yup done, thanks

**Comment by pbxqdown** [comment]
Hey guys, I'm trying to visualize the phase transition metrics. I found that duration of  phase="Failed" is dis-proportionally long compared to other phase transition duration:


This is expected since (Failed_timestamp - Running_timestamp) is the duration that VM is running which is controlled by user rather than kubevirt. 
However, I was still wondering whether there is a final phase like "End" such that the duration of cleanup/shutdown process can be calculated(e.g. shutdown_duration = End_timestamp - Failed_timestamp)
@rthallisey for viz

**Comment by rthallisey** [comment]
Hey @pbxqdown.  Failed or Succeeded phase should quantify the time a VMI object exists after the vm has exited.  So you're right that Running_timestamp should not be the value we compare against.  Can you file an issue and we'll raise it at the weekly SIG-scale meeting.

**Comment by pbxqdown** [comment]
Sure @rthallisey Let me file a issue describing the problem

**Comment by pbxqdown** [comment]
Here is the issue: https://github.com/kubevirt/kubevirt/issues/6235

**Pull request merged**
